genrl/deep/common/base_class.py

```diff
@@ -0,0 +1,164 @@
+        env: Any,
+        epochs: int = 100,
+        seed: Optional[int] = None,
+        render: bool = False,
+        device: Union[torch.device, str] = "cpu",
+        run_num: int = None,
+        save_model: str = None,
+        load_model: str = None,
+        save_interval: int = 50,
+    ):
+        self.network_type = network_type
+        self.env = env
+        self.epochs = epochs
+        self.seed = seed
+        self.render = render
+        self.run_num = run_num
+        self.save_model = save_model
+        self.load_model = load_model
+        self.save_interval = save_interval
+        self.observation_space = None
+        self.action_space = None
+
+        # Assign device
+        if "cuda" in device and torch.cuda.is_available():
+            self.device = torch.device(device)
+        else:
+            self.device = torch.device("cpu")
+
+        # Assign seed
+        if seed is not None:
+            set_seeds(seed, self.env)
+
+    def create_model(self) -> None:
+        """
+        Initialize all the policy networks in this method, and also \
+        initialize the optimizers and the buffers.
+        """
+        raise NotImplementedError()
+
+    def _update_params_before_select_action(self, timestep: int) -> None:
+        """
+        Update any parameters before selecting action like epsilon for decaying epsilon greedy
+
+        :param timestep: Timestep in the training process
+        :type timestep: int
+        """
+        pass
+
+    def select_action(self, state: np.ndarray) -> np.ndarray:
+        """
+        Selection of action
+
+        :param state: Observation state
+        :type state: int, float, ...
+        :returns: Action based on the state and epsilon value
+        :rtype: int, float, ...
+        """
+        raise NotImplementedError()
+
+    def get_loss(self) -> None:
+        """
+        Calculate and return the total loss
+        """
+        raise NotImplementedError()
+
+    def update_params(self, update_interval: int) -> None:
+        """
+        Takes the step for optimizer.
+        This internally call _get_loss().
+        """
+        raise NotImplementedError()
+
+    def get_hyperparameters(self) -> Dict[str, Any]:
+        """
+        Hyperparameters you want to return
+        """
+        raise NotImplementedError()
+
+
+class BaseOnPolicyAgent(BaseAgent):
+    def __init__(
+        self,
+        network_type: str,
+        env: Any,
+        timesteps_per_actorbatch: int = 256,
+        layers: Tuple = (64, 64),
+        gamma: float = 0.99,
+        lr_policy: float = 0.01,
+        lr_value: float = 0.0005,
+        actor_batch_size: int = 64,
+        epochs: int = 100,
+        seed: Optional[int] = None,
+        render: bool = False,
+        device: Union[torch.device, str] = "cpu",
+        run_num: int = None,
+        save_model: str = None,
+        load_model: str = None,
+        save_interval: int = 50,
+        rollout_size: int = 2048,
```

Comment:
Some can be replaced with `*args, **kwargs`

Reply:
I don't really think it'll be the correct thing here, so not doing this as of now, but if I think it might help in future then I'll add it.

Reply:
Why wouldn't it be correct?

Reply:
Yea coz we need these all params to be there, we generally use `**kwargs` for optional parameters no? (Only seed, device seem optional to me)

Reply:
No, I don't think there is anything about it being for optional parameters

Reply:
Yea but if we have parameters that are used majorly we don't generally put it in `**kwargs`, this is what I believe :|

Reply:
Why not? **kwargs will just store the parameters it cannot find in the existing keyword arguments (i.e., the ones inherited) and pass to superclasses.

Another point is it would reduce code climate issues

@ajaysub110 What are your thoughts on this?

Reply:
While **kwargs will clean stuff up, I think it will affect our docs. Its kind of annoying when the docs says args, kwargs and you're required to see the code examples to identify possible arguments to the class, no?

Reply:
Right so should keep the doctoring intact but remove arguments which can be, with args and kwargs?

Reply:
Yeah, that makes sense